File: Example/PDStrategy/PDStrategy/PDStrategy/CollectionView/PDCollectionViewCell.m
//
//  PDCollectionViewCell.m
//  PDStrategy
//
//  Created by Pavel Deminov on 19/11/2017.
//  Copyright © 2017 Pavel Deminov. All rights reserved.
//

#import "PDCollectionViewCell.h"

@interface PDCollectionViewCell ()

@end

@implementation PDCollectionViewCell

+ (NSString*)reuseIdentifier {
    NSString *classString = NSStringFromClass(self);
    classString = [[NSStringFromClass([self class]) componentsSeparatedByString:@"."] lastObject];
    return classString;
}

+ (PDCollectionViewCell *)prototype
{
    PDCollectionViewCell *prototype = [[self alloc] init];
    return prototype;
}

+ (CGSize)fitingSizeForItemInfo:(id <PDItemInfo>)itemInfo withSize:(CGSize)dynamicSize {
    PDCollectionViewCell *prototype = [self prototype];
    
    NSLayoutConstraint *width = [NSLayoutConstraint constraintWithItem:prototype.contentView attribute:(NSLayoutAttributeWidth) relatedBy:(NSLayoutRelationEqual) toItem:nil attribute:NSLayoutAttributeWidth multiplier:1.0 constant:dynamicSize.width];
    [NSLayoutConstraint activateConstraints:@[width]];
    
    prototype.itemInfo = itemInfo;
    CGSize s = [prototype systemLayoutSizeFittingSize:dynamicSize];
    
    return s;
}

- (void)setItemInfo:(id<PDItemInfo>)itemInfo {
    _itemInfo = itemInfo;
    [self updateUI];
}

- (instancetype)init {
    self = [super init];
    [self setup];
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    [self setup];
    return self;
}

- (void)updateUI {
    //override
}

- (void)setup {
    //override
}

- (void)setupForIB {
    //override
}

- (void)prepareForInterfaceBuilder {
    [self setup];
    [self setupForIB];
}

@end

File: Example/PDStrategy/PDStrategy/PDStrategy/CollectionView/PDCollectionViewCell.swift
//
//  PDCollectionViewCell.swift
//  PDStrategy
//
//  Created by Pavel Deminov on 12/11/2017.
//  Copyright © 2017 Pavel Deminov. All rights reserved.
//

import UIKit

class PDCollectionViewCell: UICollectionViewCell, PDCellInfo {
    var item: PDItemInfo? {
        didSet {
            updateUI()
        }
    }
    var reloadCellBlock: ReloadCellBlock?
    
    class var reuseIdentifier: String {
        get {
            var classString: String = NSStringFromClass(self)
            classString = (NSStringFromClass(self).components(separatedBy: ".").last) ?? "UICollectionViewCell"
            return classString
        }
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }
    
    override func prepareForInterfaceBuilder() {
        setup()
        setupForIB()
    }
    
    func setup() {
        //override
    }
    
    func setupForIB() {
        //override
    }
    
    func updateUI() {
        //override
    }
}
